src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs

```diff
@@ -1666,6 +1672,13 @@ private static HttpClient CreateHttpClient(HttpClientParameters parameters, Http
                 handler.Expect100ContinueTimeout = parameters.ContinueTimeout;
                 client.Timeout = parameters.Timeout;
 
+                if (request != null && request.ImpersonationLevel != TokenImpersonationLevel.None)
+                {
+                    // This is legacy hack. We want to deal with it only if explicitly set.
+                    var setting = typeof(SocketsHttpHandler).InvokeMember("_settings", BindingFlags.GetField | BindingFlags.NonPublic | BindingFlags.Instance, null, handler, null);
+                    setting?.GetType().InvokeMember("_impersonationLevel", BindingFlags.SetField | BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance, null, setting, new object[] { request!.ImpersonationLevel });
+                }
```

Comment:
It would be nice to structure it to be usable with `UnsafeAccessor` attribute. That's not possible here since `HttpConnectionSettings` is internal.

I assume `InvokeMember` is known reflection pattern in ILLink/ILC, right? (/cc @sbomer @MichalStrehovsky)

Reply:
`InvokeMember` is not handled. It will keep everything on `SocketsHttpHandler` because of the annotation:

https://github.com/dotnet/runtime/blob/de4fa45b047cb1f03da9d03c4d98f7518d02cc93/src/libraries/System.Private.CoreLib/src/System/Type.cs#L571-L572

It would be preferable to call `GetField` instead so that we just keep a single field.

The subsequent `setting?.GetType().InvokeMember` is not trim safe at all, this will break under trimming because we have no clue what type is this reflecting on, like the suppressed warning says. Please don't merge this like that, the warning suppression is a bug.

Reply:
updated. However, it is not clear to me how to proceed on the trimming @MichalStrehovsky. Should I annotate the `HttpConnectionSettings` somehow?

Reply:
Is `HttpConnectionSettings` the type we expect as the result of `setting?.GetType()`? If so replacing that with a `typeof` (or `Type.GetType("TypeName, AssemblyName")` if it's inaccessible) followed by `GetField("_impersonationLevel")`) should work without trimming warnings.

Reply:
Yes. But it is not public type. As @filipnavara mentioned it is `internal` to `SocketsHttpHandler`. It is generally container the handler use to pass the setting around to connection pools.

Reply:
> But it is not public type.

That doesn't matter for `Type.GetType`. You can use `Type.GetType("System.Net.Http.HttpConnectionSettings, System.Net.Http")` and it resolves the type just fine.

Reply:
updated.